core/src/main/java/org/apache/druid/java/util/common/DefineClassUtils.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.common;
+
+import org.apache.druid.utils.JvmUtils;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.security.ProtectionDomain;
+
+public class DefineClassUtils
+{
+  private static final MethodHandle DEFINE_CLASS;
+  private static final Exception DEFINE_CLASS_NOT_SUPPORTED_EXCEPTION;
+
+  static {
+    MethodHandle defineClass = null;
+    Exception exception = null;
+    try {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      if (JvmUtils.isIsJava9Compatible()) {
+        defineClass = defineClassJava9(lookup);
+      } else {
+        defineClass = defineClassJava8(lookup);
+      }
+    }
+    catch (ReflectiveOperationException | RuntimeException e) {
+      exception = e;
+    }
+    if (defineClass != null) {
+      DEFINE_CLASS = defineClass;
+      DEFINE_CLASS_NOT_SUPPORTED_EXCEPTION = null;
+    } else {
+      DEFINE_CLASS = null;
+      DEFINE_CLASS_NOT_SUPPORTED_EXCEPTION = exception;
+    }
+  }
+
+  /**
+   * "Compile" a MethodHandle that is equivalent to the following closure:
+   *
+   *  Class<?> defineClass(Class targetClass, String className, byte[] byteCode) {
+   *    MethodHandles.Lookup targetClassLookup = MethodHandles.privateLookupIn(targetClass, lookup);
+   *    return targetClassLookup.defineClass(byteCode);
+   *  }
+   */
+  private static MethodHandle defineClassJava9(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+
+    // this is getting meta
```

Comment:
Leftover comment? It's not clear what does it mean.

Reply:
"meta" here means self-referential. In this case we access method handles on the same class we also use to define method handles. This is meant as amusing editorial comment, more than anything else.

Reply:
Is it not clear? It seems pretty clear.

Reply:
the comment at the class level already implies that MethodHandles.Lookup.defineClass is only available in java 9 and above